File: examples/Objective-C/hwclient.m
//
//  Hello World client
//  Connects REQ socket to tcp://localhost:5555
//  Sends "Hello" to server, expects "World" back
//
#import "ZMQObjC.h"

int
main (void)
{
	NSAutoreleasePool *pool = [[NSAutoreleasePool alloc] init];
	ZMQContext *ctx = [[[ZMQContext alloc] initWithIOThreads:1U] autorelease];
	
	/* Get a socket to talk to clients. */
	NSLog(@"Connecting to hello world server...");
	static NSString *const kEndpoint = @"tcp://localhost:5555";
	ZMQSocket *requester = [ctx socketWithType:ZMQ_REQ];
	BOOL didBind = [requester connectToEndpoint:kEndpoint];
	if (!didBind) {
		NSLog(@"*** Failed to bind to endpoint [%@].", kEndpoint);
		return EXIT_FAILURE;
	}

	static const int kMaxRequest = 10;
	NSData *const request = [@"Hello" dataUsingEncoding:NSUTF8StringEncoding];
	for (int request_nbr = 0; request_nbr < kMaxRequest; ++request_nbr) {
		NSAutoreleasePool *localPool = [[NSAutoreleasePool alloc] init];

		NSLog(@"Sending request %d.", request_nbr);
		[requester sendData:request withFlags:0];
		NSData *reply = [requester receiveDataWithFlags:0];
		NSString *text = [[[NSString alloc]
				initWithData:reply encoding:NSUTF8StringEncoding] autorelease];
		NSLog(@"Received reply %d: %@", request_nbr, text);

		[localPool drain];
	}

	[requester close];
	[pool drain];
	return EXIT_SUCCESS;
}

File: examples/Free Pascal/hwclient.pas
{
  Hello World client
  Connects REQ socket to tcp://localhost:5555
  Sends "Hello" to server, expects "World" back
  @author cpicanco <cpicanco@ufpa.br>
}
program hwclient;

{$MODE objfpc}{$H+}

uses zmq;

var
  context, requester : Pointer;
  i : integer;
  buffer : array [0..4] of Char; 
  S : string = 'Hello'; // ansistring

begin
  WriteLn('Connecting to hello world server…');
  context := zmq_ctx_new;
  requester := zmq_socket(context, ZMQ_REQ);
  zmq_connect(requester, 'tcp://localhost:5555');
  for i := 0 to 9 do
    begin
      WriteLn('Sending '+S+'…');
      zmq_send(requester, @S[1], Length(S), 0);
      zmq_recv(requester, @buffer, Length(buffer), 0);
      WriteLn('Received ', buffer);
    end;
  zmq_close(requester);
  zmq_ctx_destroy(context);
end.
